posts/2021-05-14-paying-homage-to-the-compall-michael-representation.adoc

```diff
@@ -0,0 +1,466 @@
+rocket.distanceTravelled
+----
+
+Furthermore, if we wanted to enforce the constraints about distance only being positive, we could take the "smart constructor" approach to restrict how we obtain values of type `Kilometers` and `Miles`. There are quite a number of ways to achieve this in Scala. For a good explanation of the various ways to implement smart constructors I invite you to check out https://tuleism.github.io/blog/2020/scala-smart-constructors/[this recent blog post]. As I am using Scala 2.13.5 to compile these examples, I will take advantage of the newer cleaner approach of using a `private[foo]` access modifier that is possible with the `-Xsource:3` compiler option. (If you are on a version of Scala pre-2.13.2, then I would personally recommend the approach of using `sealed abstract case class`, which is also explained in the linked blog post).
+
+In the following code snippet, the `private[Units]` modifier means we can only access the `Kilometers()` and `Miles()` constructors from within the scope of `object Units` -- externally the only way to obtain `Kilometers` and `Miles` values is through the `kilometres` and `miles` "smart constructors" that perform some validation.
+
+[source,scala]
+----
+object Units {
+  final case class Kilometers private[Units] (value: Double)
+  final case class Miles private[Units] (value: Double)
+
+  val ZeroKm: Kilometers = Kilometers(0)
+  val ZeroMi: Miles = Miles(0)
+
+  def kilometres(value: Double): Option[Kilometers] = if (value < 0) None else Some(Kilometers(value))
+  def miles(value: Double): Option[Miles] = if (value < 0) None else Some(Miles(value))
+
+  def add(km1: Kilometers, km2: Kilometers): Kilometers = Kilometers(km1.value + km2.value)
+  def toKilometers(miles: Miles): Kilometers = Kilometers(miles.value * 1.6)
+}
+
+import Units._
+
+class Booster() {
+  def provideLaunchBoost(): Miles = miles(100).getOrElse(ZeroMi)
+}
+
+class Rocket(booster: Booster) {
+  private var distance: Kilometers = ZeroKm
+
+  def launch(): Unit = {
+    // Kilometers and Miles are different types. So compiler prevents the previous bug
+    val launchBoost: Kilometers = toKilometers(booster.provideLaunchBoost())
+    distance = add(distance, launchBoost)
+  }
+
+  def distanceTravelled: Kilometers = distance
+}
+
+val rocket: Rocket = new Rocket(new Booster())
+rocket.launch();
+
+// Will represent the correct distance travelled
+rocket.distanceTravelled
+----
+
+So we have achieved what we wanted -- which is that we prevent the bug at compile-time -- but at the cost of some runtime overhead because we now have to allocate wrapper `Kilometers` and `Miles` objects in memory. It is safe to bet that in practice for most applications this extra overhead is not critical and personally I reach for this simple approach most of the times. However, it is not unheard of to have some performance-critical use-cases where you want to avoid the overhead of unnecessarily allocating and garbage-collecting objects.
```

Comment:
```suggestion
So we have achieved what we wanted -- which is that we prevent the bug at compile-time -- but at the cost of some runtime overhead because we now have to allocate the wrapper `Kilometers` and `Miles` objects in memory. It is safe to bet that in practice for most applications this extra overhead is not critical and personally I reach for this simple approach most of the times. However, it is not unheard of to have some performance-critical use-cases where you want to avoid the overhead of unnecessarily allocating and garbage-collecting objects.
```